akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -394,6 +396,65 @@ final case class Fold[In, Out](zero: Out, f: (Out, In) ⇒ Out) extends GraphSta
 /**
  * INTERNAL API
  */
+final class FoldAsync[In, Out](zero: Out, f: (Out, In) ⇒ Future[Out]) extends GraphStage[FlowShape[In, Out]] {
+  import akka.dispatch.ExecutionContexts
+
+  val in = Inlet[In]("FoldAsync.in")
+  val out = Outlet[Out]("FoldAsync.out")
+  val shape = FlowShape.of(in, out)
+
+  override def toString: String = "FoldAsync"
+
+  override val initialAttributes = DefaultAttributes.foldAsync
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new SupervisedGraphStageLogic(inheritedAttributes, shape) with InHandler with OutHandler {
+      private var aggregator: Out = zero
+      private var aggregating: Future[Out] = Future.successful(aggregator)
+
+      override def onResume(t: Throwable): Unit = {
+        aggregator = zero
+      }
+
+      private def handle(output: Try[Out]): Unit = output match {
+        case Failure(reason) ⇒ fail(out, reason)
+        case Success(o) ⇒
+          aggregator = o
+
+          if (!isClosed(in)) pull(in)
+          else completeStage()
```

Comment:
Where you can control what state your stage is in when the future completes. The probes allow you to control this on a very low level. Look at `FlowMapAsyncSpec` for example.